File: src/main/java/com/EmosewaPixel/pixellib/materialSystem/lists/Materials.java
package com.EmosewaPixel.pixellib.materialsystem.lists;

import com.EmosewaPixel.pixellib.materialsystem.materials.Material;

import java.util.Collection;
import java.util.HashMap;

//This class contains functions for interacting with the global list of materials
public final class Materials {
    private static final HashMap<String, Material> materials = new HashMap<>();

    public static void add(Material mat) {
        if (materials.get(mat.getName()) != null)
            materials.get(mat.getName()).merge(mat);
        else
            materials.put(mat.getName(), mat);
    }

    public static Material get(String name) {
        return materials.get(name);
    }

    public static boolean exists(String name) {
        return get(name) != null;
    }

    public static Collection<Material> getAll() {
        return materials.values();
    }
}

File: src/main/java/com/emosewapixel/pixellib/materialsystem/lists/Materials.kt
package com.emosewapixel.pixellib.materialsystem.lists

import com.blamejared.crafttweaker.api.annotations.ZenRegister
import com.emosewapixel.pixellib.materialsystem.main.Material
import org.openzen.zencode.java.ZenCodeGlobals
import org.openzen.zencode.java.ZenCodeType

//This class contains functions for interacting with the global list of materials
@ZenRegister
@ZenCodeType.Name("pixellib.materialsystem.lists.Materials")
object Materials {
    private val materials = hashMapOf<String, Material>()

    @ZenCodeGlobals.Global("materials")
    val instance = this

    @JvmStatic
    val all
        @ZenCodeType.Getter get() = materials.values.distinct()

    @JvmStatic
    fun add(mat: Material) {
        val merged = mat.names.mapNotNull(this::get).distinct().fold(mat) { acc, current -> current.merge(acc) }
        mat.names.forEach { materials[it] = merged }
    }

    @JvmStatic
    @ZenCodeType.Operator(ZenCodeType.OperatorType.INDEXGET)
    operator fun get(name: String) = materials[name]

    @JvmStatic
    @ZenCodeType.Operator(ZenCodeType.OperatorType.CONTAINS)
    operator fun contains(name: String) = name in materials
}